packages/core-mobile/app/new/features/meld/store.ts

```diff
@@ -57,3 +62,105 @@ export const useOfframpActivityIndicator = (): {
 } => {
   return offrampActivityIndicatorStore()
 }
+
+type TokenIndex = {
+  nativeMap: Map<string, LocalTokenWithBalance>
+  erc20Map: Map<string, LocalTokenWithBalance>
+  btcToken?: LocalTokenWithBalance
+}
+
+type TokenIndexStore = {
+  tokenIndex: TokenIndex | null
+  setTokenIndex: (tokens: LocalTokenWithBalance[]) => void
+  reset: () => void
+}
+
+const tokenIndexStore = create<TokenIndexStore>((set, get) => ({
+  tokenIndex: null,
+  setTokenIndex: tokens => {
+    if (get().tokenIndex) return // already built, skip
+
+    const nativeMap = new Map<string, LocalTokenWithBalance>()
+    const erc20Map = new Map<string, LocalTokenWithBalance>()
+    let btcToken: LocalTokenWithBalance | undefined
+
+    for (const token of tokens) {
+      if (token.type === 'NATIVE') {
+        nativeMap.set(token.networkChainId.toString(), token)
+      }
+      if ('chainId' in token && token.address) {
+        erc20Map.set(`${token.chainId}-${token.address.toLowerCase()}`, token)
+      }
+      if (token.symbol === 'BTC') {
+        btcToken = token
+      }
+    }
+
+    set({ tokenIndex: { nativeMap, erc20Map, btcToken } })
+  },
+  reset: () => set({ tokenIndex: null })
+}))
+
+export const useTokenIndex = (): TokenIndexStore => {
+  return tokenIndexStore()
+}
+
+type SupportedCryptoCurrenciesStore = {
+  supportedCryptoCurrencies: CryptoCurrencyWithBalance[]
+  setSupportedCryptoCurrencies: (
+    cryptos: CryptoCurrency[],
+    tokenIndexes: TokenIndex
+  ) => void
+  reset: () => void
+}
+
+const supportedCryptoCurrenciesStore = create<SupportedCryptoCurrenciesStore>(
```

Comment:
also caching the data in memory/zustand store so that every time user opens the select a token screen, it doesn't recompute. this data will get reset whenever the buy/sell flow starts so a new list can get computed.